scripts/calculateRevenue/protocols/utils/networks.ts

```diff
@@ -0,0 +1,59 @@
+import { QueryResponse, TransactionField } from '@envio-dev/hypersync-client'
+import { getHyperSyncClient } from '../../../utils'
+import { NetworkId } from '../../../types'
+
+export async function fetchTotalGasUsed({
+  networkId,
+  users,
+  startBlock,
+  endBlock,
+}: {
+  networkId: NetworkId
+  users: string[]
+  startBlock?: number
+  endBlock?: number
+}): Promise<number> {
+  let fromBlock = startBlock ?? 0
+  let totalGasUsed = 0
+  let hasMoreBlocks = true
+
+  try {
+    const client = getHyperSyncClient(networkId)
+
+    const query = {
+      transactions: [{ from: users }],
+      fieldSelection: {
+        transaction: [TransactionField.GasUsed, TransactionField.GasPrice],
+      },
+      fromBlock,
+      ...(endBlock && { toBlock: endBlock }),
+    }
+
+    do {
+      const response: QueryResponse = await client.get(query)
+
+      if (
+        response.nextBlock <= fromBlock ||
+        !response.data.transactions.length
```

Comment:
How does pagination work for this endpoint? Is the page size equal to a fixed number of blocks, or a fixed number of TXs? In other words, is it possible for the endpoint to return 0 transactions in a response? Like if we requested block range 0-1000, and there were no TXs between 0-100 (but transactions afterwards), would the query return 0 TXs the first time? If so, the `!response.data.transactions.length` condition would cause `hasMoreBlocks` to be set prematurely.

Reply:
Good catch! Pagination constraints can be set either via `maxNumBlocks` or `maxNumTransactions`; however, the default numbers for these values [were not found](https://github.com/enviodev/hypersync-client-node/blob/ef55e364d26845c612a5233fdf8a2c182b2ffefa/index.d.ts#L242-L312). In testing are there are cases where transactions return an empty array with blocks were remaining to be processed. I'll remove this check.